scrapy/exporters.py

```diff
@@ -252,13 +258,21 @@ def _build_row(self, values):
 
     def _write_headers_and_set_fields_to_export(self, item):
         if self.include_headers_line:
+            if isinstance(item, dict):
+                # for dicts try using fields of the first item
+                fields = list(item.keys())
+            else:
+                # use fields declared in Item
+                fields = list(item.fields.keys())
             if not self.fields_to_export:
-                if isinstance(item, dict):
-                    # for dicts try using fields of the first item
-                    self.fields_to_export = list(item.keys())
-                else:
-                    # use fields declared in Item
-                    self.fields_to_export = list(item.fields.keys())
+                self._is_data_loss_warning_displayed = True
```

Comment:
I'm failing to see the need for two flags to handle this feature. At the very least, I don't understand the names, i.e. how `self.fields_to_export` being unset implies setting `self._is_data_loss_warning_displayed = True`.
Could this be simplified to use only one flag, or maybe do some changes to have more self-explanatory flag names?

Reply:
There are three cases when we should show the warning:
1. When the `fields_to_export ` is not set and we get the fields from the first row.
2. When the `fields_to_export ` is set but the fields in the rows differ
3. When the `fields_to_export ` is not set and we get the fields from the first row but the subsequent rows differ in content from the first row.

So simply checking the `fields_to_export ` is set or not won't let me understand whether to show the warning or not so i need a flag to check the conditions 2 & 3.

Also I need another flag to check if I have already shown the warning, if it is shown already then I should not show it again.

I renamed the flags to `_show_data_loss_warning` & ` _data_loss_already_warned`.

Reply:
@elacuesta did I get it right? what do you think? :smiley:

Reply:
I believe the following piece of code should be equivalent:
```python
if self.fields_to_export:
    if set(fields) != set(self.fields_to_export) and not self._data_loss_already_warned:
        logger.warning("Data loss detected when exporting items -- This message won't be shown in further items")
        self._data_loss_warned = True
else:
    self.fields_to_export = fields
```
Tests fail because this ignores (1), which I think should not be accounted for (imagine `fields_to_export` is unset but all fields have the same fields).

However, while checking this I realized the fix only runs on `_write_headers_and_set_fields_to_export`, which is called only once (if the header row was not written already). If I understand the issue correctly, we would want this check to be executed for every item.

Reply:
You mean apart from adding the fix in `_write_headers_and_set_fields_to_export` we need to add it in `export_item` method also?

Reply:
Then we can define a new attribute called `_fields_size` and set to an initial value, now if it varies because of the changing row size we would trigger the warning?

Reply:
@elacuesta I added a new commit to check if data loss is happening for every item, the check is added in the `export_item ` method.